### PR TITLE
Branch support in resolver

### PR DESCRIPTION
**Pull request by aciidb0mb3r**

### CODE REVIEW

File: Sources/PackageGraph/DependencyResolver.swift [hunk 1]
```diff
@@ -151,6 +198,18 @@ public protocol PackageContainer {
...
+    /// Fetch the declared dependencies for a particular revision.
+    ///
+    /// This property is expected to be efficient to access, and cached by the
+    /// client if necessary.
+    ///
+    /// - Throws: If the revision could not be resolved; this will abort
+    ///   dependency resolution completely.
+    //
+    // FIXME: Should we use just use the `Revision` structure from
+    // SourceControl here?
```
Comment (ddunbar) [inline Sources/PackageGraph/DependencyResolver.swift]: I dunno, its pretty nice to keep it decoupled.

File: Sources/PackageGraph/DependencyResolver.swift [hunk 2]
```diff
@@ -171,6 +230,9 @@ public struct PackageContainerConstraint<T: PackageContainerIdentifier>: CustomS
         /// The requirement is specified by the version set.
         case versionSet(VersionSetSpecifier)
 
+        /// The requirement is specified by the revision.
+        case revision(String)
```
Comment (ddunbar) [inline Sources/PackageGraph/DependencyResolver.swift]: This would be the place to document exactly what this means -- that it is an arbitrary string which must match and be totally equal to be valid. I think there is a probably a good generalization of this concept, but I don't see a need to tackle it now. That said, let's completely document the behavior.

File: Sources/PackageGraph/DependencyResolver.swift [hunk 3]
```diff
@@ -531,6 +634,17 @@ struct VersionAssignmentSet<C: PackageContainer>: Equatable, Sequence {
...
 
+        case .revision(let identifier):
+            // If we already have a revision constraint, it should be same as the one
+            // we're trying to set.
+            if case .revision(let existingRevision) = constraints[container.identifier] {
+                return existingRevision == identifier
+            }
+            // Otherwise, it is always valid to set a revision binding.  Note
+            // that there are rules that prevents versioned constaints from
+            // having revision constaints, but that is handled by the resolver.
```
Comment (ddunbar) [inline Sources/PackageGraph/DependencyResolver.swift]: s/constaints/constraints/

Comment (ddunbar) [inline Sources/PackageGraph/DependencyResolver.swift]: I'm not sure I understand this -- why shouldn't this return false if there is an existing version constraint; this seems to violate the API contract?

File: Sources/PackageGraph/DependencyResolver.swift [hunk 4]
```diff
@@ -787,6 +907,24 @@ public class DependencyResolver<
...
+                    }
+                    return nil
+                }
+                // If we have any revision constraints, set the error and abort.
+                guard revisionConstraints.isEmpty else {
+                    self.error = DependencyResolverError.revisionConstraints(
+                        dependency: AnyPackageContainerIdentifier(container.identifier),
+                        revisions: revisionConstraints)
+                    return AnySequence([])
+                }
```
Comment (ddunbar) [inline Sources/PackageGraph/DependencyResolver.swift]: This doesn't _feel_ right to me -- I would have expected this to be a natural side effect of other parts of the code. I expect this to return the versions which are *immediately* valid but the filter based on other constraints in the system to happen in response to attempts to merge the binding into the assignment. Does that not work?

File: Sources/PackageGraph/RepositoryPackageContainerProvider.swift [hunk 5]
```diff
@@ -161,15 +161,23 @@ public class RepositoryPackageContainer: PackageContainer, CustomStringConvertib
...
+        let revision = try repository.resolveRevision(tag: tag)
+        return try getDependencies(at: revision, version: version)
+    }
+
+    public func getDependencies(at revision: String) throws -> [RepositoryPackageConstraint] {
+        return try getDependencies(at: Revision(identifier: revision))
+    }
+
+    /// Returns dependencies of a container at the given revision.
+    private func getDependencies(at revision: Revision, version: Version? = nil) throws -> [RepositoryPackageConstraint] {
```
Comment (ddunbar) [inline Sources/PackageGraph/RepositoryPackageContainerProvider.swift]: 👍 -- FWIW I would probably land something like this in its own patch just to separate obvious infra improvement from hard algorithm things

File: Sources/PackageGraph/DependencyResolver.swift (hunk 3 above)
Comment (aciidb0mb3r) [inline Sources/PackageGraph/DependencyResolver.swift]: If A is versioned transitively and if root package wants to use A on a branch, that should work, right?

File: Sources/PackageGraph/RepositoryPackageContainerProvider.swift (hunk 5 above)
Comment (aciidb0mb3r) [inline Sources/PackageGraph/RepositoryPackageContainerProvider.swift]: Yeah, its a sep commit, I'll land it separately.

Comment (ddunbar) [inline Sources/PackageGraph/RepositoryPackageContainerProvider.swift]: Oh -- I didn't notice that; yeah never mind this is perfect. :)
